### PR TITLE
TODO: Provider - Device code

### PR DESCRIPTION
**Pull request by jcampbell05**
This PR focuses on the server implementation of #626 and leaves the client implementation for a future PR. Since its in a draft I have placed it in a draft folder to indicate this so those who need it can use it with the awareness it may change in the future and will be re-named when the specification is stable.

Device Code is a new authorization flow which allows a client to be authenticated by a user for a device with limited or no input. 

I based this implementation on the authorization grant type since this flow shares some similarities but I utilized some concepts for the client grant type where this flow differs.

To understand how to implement the new classes and flow for a server here is an overflow of the flow and how a server would use them:

1. The server implements an authorization endpoint the client can use to request a user & device code. When this endpoint is hit the DeviceCodeGrant's `create_authorization_response` is called. A DeviceToken should be passed in with the expiry date configured by the server as well as a RequestValidator configured to return the verification url and polling interval. 
2. The client displays the user code to the user and the verification url. Once user enters code which exists . The server calls DeviceCodeGrant's `create_verification_response` which uses the validator to check that code exists.
3. If it exists then the server will ask user to allow or disallow. The server calls DeviceCodeGrant's `create_verification_response` which will this time ask the validator to update the authorized state.
4. Whilst this happening the client is polling a endpoint for the token status. Each time the server will call DeviceCodeGrant's `create_token_response` method. The validator's validate_token_request will validate the device token is authorized and should implement all of the error codes from the RFC (including rate limiting if needed).

Once authorized this method will use the create_token method on the BearerToken to create a new Access Token, ask the validator to save it and then invalidate the old device code.

### CODE REVIEW

**Comment by duaneking** [comment]
This PR will be deleted/closed if the issues are not fixed.

I'm leery of a custom login flow that doesn't cite any RFC;s being added to the lib anyway, since it's a new attack vector.

**Comment by jcampbell05** [comment]
> This PR can not be included or merged if the issues are not fixed.
> 

Of course but I'm still working on it locally in a project to test the APIs.

> I'm leery of a custom login flow that doesn't cite any RFC;s being added to the lib anyway, since it's a new attack vector.
> 
> What RFC's does this implement? This seems like a custom flow and as a one-off pr for an unknown process flow with no RFC or standardization, I'm inclined to reject it anyway.
> 

It implements this draft flow https://tools.ietf.org/html/draft-ietf-oauth-device-flow-13

> Edit: I looked at the original link and I'm not sure this is a good fit for the lib until we support JWT's fully for clients.

This flow doesn't require JWT so not sure what the blocker is there.

**Comment by duaneking** [comment]
Its architectural.

If you look at the token types and the generator it looks like some test shims exist for JWT that point out the problems in the WebApp client, but they are otherwise dead code as thye are only referenced in tests.

For example, grant types etc are currently assumed, so that architectural work will need to be done to support multiple, in order to support both this device stuff and the JWT.

**Comment by duaneking** [comment]
The change hasn't happened yet, its just something I noticed the other day so I'm thinking we should work to common goals.  The grant_type for example is hard coded but for JTW you need a very different value because an assert needs to happen that the value was used.

We need to be able to support a different type based on the type of token being used; if you want to do that work as part of your PR it would help with support for both.

**Comment by thedrow** [comment]
FYI we're at [draft 15](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15) now.
If you still want to work on the PR, you'll have to rebase and update it.

**Comment by jcampbell05** [comment]
@thedrow no worries, I will do so.

**Comment by JonathanHuot** [comment]
Finalized as RFC8628 : https://tools.ietf.org/html/rfc8628

**Comment by jcampbell05** [comment]
I'l restart this work with the final RFC :)

**Comment by thedrow** [comment]
@jcampbell05 Ping?
Any change to get some progression on this PR soon or should we close it for now?

**Comment by thedrow** [comment]
@jcampbell05 Is that a no?
Last chance before we close this and open an issue instead.

**Comment by jcampbell05** [comment]
No lets close it and open an issue - I've ran out of bandwidth at the moment but this can server as a POC at least for anyone wanting to pick this back up

**Comment by auvipy** [comment]
> No lets close it and open an issue - I've ran out of bandwidth at the moment but this can serve as a POC at least for anyone wanting to pick this back up
> 
> If I have time in the future then can open a new PR

I will push on your branch. I will take on this :)

**Comment by auvipy** [comment]
> @auvipy Is this still being actively worked on?

no, need to come back to this, are you interested?

File: oauthlib/oauth2/draft_ietf13/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+"""
+oauthlib.oauth2.draft_ietf13
+~~~~~~~~~~~~~~~~~~~~~~~
+
+This module is an implementation of various logic needed
+for consuming and providing OAuth 2.0 IETF13 Draft.
+"""
+
+from __future__ import absolute_import, unicode_literals
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/__init__.py:10]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/endpoints/code.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+"""
+oauthlib.oauth2.draft_ietf13
+~~~~~~~~~~~~~~~~~~~~~~~
+
+This module is an implementation of various logic needed
+for consuming and providing OAuth 2.0 Draft IETF 13.
+"""
+from __future__ import absolute_import, unicode_literals
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/endpoints/code.py:9]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/endpoints/verification.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/endpoints/verification.py:1]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/endpoints/verification.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+"""
+oauthlib.oauth2.draft_ietf13
+~~~~~~~~~~~~~~~~~~~~~~~
+
+This module is an implementation of various logic needed
+for consuming and providing OAuth 2.0 Draft IETF 13.
+"""
+from __future__ import absolute_import, unicode_literals
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/endpoints/verification.py:9]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/grant_types/__init__.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# -*- coding: utf-8 -*-
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/grant_types/__init__.py:1]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/grant_types/__init__.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+"""
+oauthlib.oauth2.draft_ietf13.grant_types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""
+from __future__ import unicode_literals, absolute_import
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/grant_types/__init__.py:6]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/grant_types/device_code.py [hunk 7]
```diff
@@ -0,0 +1,562 @@
+# -*- coding: utf-8 -*-
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/grant_types/device_code.py:1]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/grant_types/device_code.py [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+"""
+oauthlib.oauth2.draft_ietf13.grant_types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""
+from __future__ import absolute_import, unicode_literals
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/grant_types/device_code.py:6]:
```suggestion
```

File: oauthlib/oauth2/draft_ietf13/tokens.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
+"""
+oauthlib.oauth2.draft_ietf13.tokens
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This module contains methods for adding device and user codes to requests.
+
+- Bearer https://tools.ietf.org/html/rfc6750
+"""
+from __future__ import absolute_import, unicode_literals
```
Comment (auvipy) [inline oauthlib/oauth2/draft_ietf13/tokens.py:9]:
```suggestion
```

File: oauthlib/oauth2/rfc6749/endpoints/introspect.py [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
```
Comment (auvipy) [inline oauthlib/oauth2/rfc6749/endpoints/introspect.py:1]:
```suggestion
```

File: oauthlib/oauth2/rfc6749/endpoints/pre_configured.py [hunk 11]
```diff
@@ -1,32 +1,36 @@
 # -*- coding: utf-8 -*-
```
Comment (auvipy) [inline oauthlib/oauth2/rfc6749/endpoints/pre_configured.py:1]:
```suggestion
```

File: oauthlib/oauth2/rfc6749/endpoints/resource.py [hunk 12]
```diff
@@ -1,6 +1,6 @@
 # -*- coding: utf-8 -*-
```
Comment (auvipy) [inline oauthlib/oauth2/rfc6749/endpoints/resource.py:1]:
```suggestion
```

File: tests/oauth2/draft_ietf13/grant_types/test_device_code.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
```
Comment (auvipy) [inline tests/oauth2/draft_ietf13/grant_types/test_device_code.py:1]:
```suggestion
```

File: tests/oauth2/draft_ietf13/grant_types/test_device_code.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+from __future__ import absolute_import, unicode_literals
```
Comment (auvipy) [inline tests/oauth2/draft_ietf13/grant_types/test_device_code.py:2]:
```suggestion
```

**Review by auvipy — CHANGES_REQUESTED** [review]

**Comment by brianhelba** [comment]
@jcampbell05 @auvipy Is this obsolete now that #881 is merged, or does this PR provide additional features beyond that?

**Comment by auvipy** [comment]
I think it is superseded by new ones

**Comment by jcampbell05** [comment]
> @jcampbell05 @auvipy Is this obsolete now that #881 is merged, or does this PR provide additional features beyond that?

I'm not familiar with the other PR but I'll close this for now. If anything is needed for the final spec, future implementors can be linked to this.

**Pull request closed without merging**